# Exploring Ebay Car Sales Data
This dataset is about used cars from *eBay Kleinanziegen*, a classified section of the German eBay website. [The original dataset.](https://data.world/data-society/used-cars-data)

The aim of this project is to clean data and analyze the included used car listings.

In [3]:
import numpy as np
import pandas as pd

autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [4]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

We can see that the fields in the dataset are filled in German. Also, many values are missing.

### Cleaning Column Names

I'll convert the column names from *CamelCase* to *snake_case* and reword some column names based on the data dictionary to be more descriptive.

In [5]:
array_columns = autos.columns.copy()
print(array_columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [6]:
array_columns = array_columns.str.replace('dateCrawled', 'date_crawled')
array_columns = array_columns.str.replace('offerType', 'offer_type')
array_columns = array_columns.str.replace('vehicleType', 'vehicle_type')
array_columns = array_columns.str.replace('yearOfRegistration', 'registration_year')
array_columns = array_columns.str.replace('powerPS', 'power_ps')
array_columns = array_columns.str.replace('monthOfRegistration', 'registration_month')
array_columns = array_columns.str.replace('fuelType', 'fuel_type')
array_columns = array_columns.str.replace('notRepairedDamage', 'unrepaired_damage')
array_columns = array_columns.str.replace('dateCreated', 'ad_created')
array_columns = array_columns.str.replace('nrOfPictures', 'number_of_pics')
array_columns = array_columns.str.replace('postalCode', 'postal_code')
array_columns = array_columns.str.replace('lastSeen', 'last_seen')

print(array_columns)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'number_of_pics', 'postal_code',
       'last_seen'],
      dtype='object')


In [7]:
autos.columns = array_columns
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  number_of_pics  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

### Initial Exploration and Cleaning
* Looking for text columns with the same values, which can be dropped as they don't have useful info for analysis
* Cleaning and converting numeric data stored as text

In [8]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-08 10:40:35  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  number_of_pics   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

* So, the columns `seller` and `offer_type` contain almost one same value so it can be dropped.
* `price` and `odometer` contain numeric data stored as text, need to be cleaned


In [9]:
autos.drop(columns=['seller', 'offer_type', 'number_of_pics'], inplace=True)

In [10]:
autos['vehicle_type'].value_counts()





limousine     12859
kleinwagen    10822
kombi          9127
bus            4093
cabrio         3061
coupe          2537
suv            1986
andere          420
Name: vehicle_type, dtype: int64

In [11]:
autos['odometer'].value_counts()

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64

In [12]:
autos['fuel_type'].value_counts()

benzin     30107
diesel     14567
lpg          691
cng           75
hybrid        37
andere        22
elektro       19
Name: fuel_type, dtype: int64

In [13]:
autos['unrepaired_damage'].value_counts()

nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64

In [14]:
autos['registration_month'].head()

0    3
1    6
2    7
3    6
4    7
Name: registration_month, dtype: int64

Converting `price` and `odometer` to number

In [15]:
autos['price'] = autos['price'].str.replace('$', '').str.replace(',', '').astype(int)
autos.rename(columns={'price' : 'price_$'}, inplace=True)
autos.head(1)

date_crawled                              name  price_$   abtest  \
0  2016-03-26 17:47:46  Peugeot_807_160_NAVTECH_ON_BOARD     5000  control   

  vehicle_type  registration_year  gearbox  power_ps   model   odometer  \
0          bus               2004  manuell       158  andere  150,000km   

   registration_month fuel_type    brand unrepaired_damage  \
0                   3       lpg  peugeot              nein   

            ad_created  postal_code            last_seen  
0  2016-03-26 00:00:00        79588  2016-04-06 06:45:54

In [16]:
autos['odometer'] = autos['odometer'].str.replace(',', '').str.replace('km', '').astype(int)
autos.rename(columns={'odometer' : 'odometer_km'}, inplace=True)
autos.head(1)

date_crawled                              name  price_$   abtest  \
0  2016-03-26 17:47:46  Peugeot_807_160_NAVTECH_ON_BOARD     5000  control   

  vehicle_type  registration_year  gearbox  power_ps   model  odometer_km  \
0          bus               2004  manuell       158  andere       150000   

   registration_month fuel_type    brand unrepaired_damage  \
0                   3       lpg  peugeot              nein   

            ad_created  postal_code            last_seen  
0  2016-03-26 00:00:00        79588  2016-04-06 06:45:54

### Exploring the Odometer and Price Columns

Nextly, specifically looking for unrealistic values in `price_$` and `odometer_km` to remove or change them

In [17]:
print(autos['odometer_km'].unique().shape)
print(autos['odometer_km'].describe())

(13,)
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


Everything is okay with odometer's min and max values.

In [18]:
print(autos['price_$'].unique().shape)
print(autos['price_$'].describe())
print(autos['price_$'].value_counts().sort_index().head(10))
print(autos['price_$'].value_counts().sort_index(ascending=False).head(10))


(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_$, dtype: float64
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
Name: price_$, dtype: int64
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price_$, dtype: int64


Many price values are listed as 0, which most likely meant that the price was not indicated. Also, I will remove all values up to $ 500, because the price is unrealistically understated, and, perhaps, this price was indicated as an lease, and not a sale.

As for the upper border, we see several overpriced positions. The most expensive car on the market on 2016 is [$4,8 miilion](https://groco.com/article/top-8-most-expensive-cars-in-2016/)

We'll reduce the number of rows by the specified value boundaries

In [19]:
autos = autos[autos['price_$'].between(500, 4800000)]
autos['price_$'].describe()

count    4.510300e+04
mean     6.528775e+03
std      2.358768e+04
min      5.000000e+02
25%      1.500000e+03
50%      3.500000e+03
75%      7.900000e+03
max      3.890000e+06
Name: price_$, dtype: float64

5000 inaccurate values are gone, minimum and maximum are fine now.

### Exploring the date columns

Now let's move on to the date columns (there are 5) and understand the date range the data covers. 

In [20]:
print(autos['date_crawled'].value_counts(dropna=False).sort_index().head())
print('...\n',autos['date_crawled'].value_counts(dropna=False)
      .sort_index().tail()
     )

2016-03-05 14:06:30    1
2016-03-05 14:06:40    1
2016-03-05 14:07:08    1
2016-03-05 14:07:21    1
2016-03-05 14:07:26    1
Name: date_crawled, dtype: int64
...
 2016-04-07 14:30:09    1
2016-04-07 14:30:26    1
2016-04-07 14:36:44    1
2016-04-07 14:36:55    1
2016-04-07 14:36:56    1
Name: date_crawled, dtype: int64


In [21]:
print(autos['ad_created'].value_counts(dropna=False).sort_index().head())
print('...\n',autos['ad_created'].value_counts(dropna=False)
      .sort_index().tail()
     )

2015-06-11 00:00:00    1
2015-08-10 00:00:00    1
2015-09-09 00:00:00    1
2015-11-10 00:00:00    1
2015-12-05 00:00:00    1
Name: ad_created, dtype: int64
...
 2016-04-03 00:00:00    1761
2016-04-04 00:00:00    1670
2016-04-05 00:00:00     537
2016-04-06 00:00:00     147
2016-04-07 00:00:00      54
Name: ad_created, dtype: int64


In [22]:
print(autos['last_seen'].value_counts(dropna=False).sort_index().head())
print('...\n',autos['last_seen'].value_counts(dropna=False)
      .sort_index().tail()
     )

2016-03-05 14:45:46    1
2016-03-05 14:46:02    1
2016-03-05 14:49:34    1
2016-03-05 15:16:11    1
2016-03-05 15:16:47    1
Name: last_seen, dtype: int64
...
 2016-04-07 14:58:44    3
2016-04-07 14:58:45    1
2016-04-07 14:58:46    1
2016-04-07 14:58:48    3
2016-04-07 14:58:50    3
Name: last_seen, dtype: int64


`date_crawled` and `last_seen` affect the same time period lasting a month. And `ad_created` displays several months earlier.

In [23]:
autos['registration_year'].describe()

count    45103.000000
mean      2005.062878
std         89.646202
min       1000.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

So, registration year column contains wrong values like 1000 year or 9999 year. This should be removed.

### Dealing with Incorrect Registration Year Data

The ending year of registration is 2016, when a note was made. The starting year let's take as 1900, not earlier for sure.

In [24]:
test_set = autos[autos['registration_year'].between(1900, 2016)]
test_set['registration_year'].value_counts(normalize=True)

2005    0.066099
2000    0.062660
2004    0.061922
2006    0.061552
2003    0.061322
1999    0.059383
2001    0.058160
2002    0.055829
2007    0.052390
2008    0.050913
2009    0.047982
1998    0.046759
2011    0.037319
2010    0.036604
1997    0.034896
2012    0.030165
1996    0.024741
2016    0.021925
1995    0.019848
2013    0.018394
2014    0.015048
1994    0.011540
2015    0.008401
1993    0.007847
1992    0.007062
1991    0.006762
1990    0.006324
1989    0.003577
1988    0.002977
1985    0.001985
          ...   
1966    0.000485
1960    0.000439
1969    0.000439
1975    0.000415
1965    0.000392
1964    0.000254
1963    0.000185
1959    0.000138
1961    0.000138
1958    0.000092
1956    0.000092
1962    0.000092
1937    0.000092
1951    0.000046
1955    0.000046
1954    0.000046
1910    0.000046
1941    0.000046
1957    0.000046
1934    0.000046
1953    0.000023
1950    0.000023
1927    0.000023
1929    0.000023
1931    0.000023
1948    0.000023
1938    0.000023
1939    0.0000

The majority of cars were registered between ~1996 and ~2012 years.

### Exploring Price by Brand

In [25]:
brands = autos['brand'].value_counts(normalize=True)
print(brands)

volkswagen        0.213644
bmw               0.113363
opel              0.101590
mercedes_benz     0.101501
audi              0.089684
ford              0.064896
renault           0.044676
peugeot           0.028956
fiat              0.023990
seat              0.018580
skoda             0.016784
smart             0.015232
mazda             0.015054
nissan            0.015032
citroen           0.014256
toyota            0.013281
hyundai           0.010066
sonstige_autos    0.009645
volvo             0.009246
mini              0.009201
honda             0.007937
mitsubishi        0.007694
kia               0.007339
alfa_romeo        0.006563
porsche           0.006186
chevrolet         0.005986
suzuki            0.005853
chrysler          0.003570
dacia             0.002860
jeep              0.002395
land_rover        0.002195
daihatsu          0.002173
subaru            0.001973
saab              0.001641
jaguar            0.001574
daewoo            0.001352
rover             0.001264
t

We will select several brands based on the amount of data. Let's take the occurrence of more than 5%: six brands from `volkswagen` to `ford`. This choice will allow us to take the largest sample to achieve greater statistical accuracy.

In [26]:
picked_brands = []
i = 0

for el in brands:
    if el > 0.05:
        picked_brands.append(brands.index[i])
    i += 1
    
print(picked_brands)

['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford']


In [27]:
mean_price_by_brand = {}

for el in picked_brands:
    mean_price = autos[autos['brand'] == el]['price_$'].mean()
    mean_price_by_brand[el] = mean_price
    
print(mean_price_by_brand)

{'opel': 3348.936054124836, 'audi': 9484.499876390606, 'mercedes_benz': 8670.394058540847, 'bmw': 8754.601994914923, 'volkswagen': 5907.612494811125, 'ford': 4607.311923471131}


Here we can see that Audi, BMW and Mercedes Benz have the most expensive average cost, because they are premial brands. The cheapest is Opel, and the medium are Volkswagen and Ford.

### Storing Aggregate Data in a DataFrame

Also, let's calculate the mean mileage the same way and then create a separate dataframe (table) with these values.

In [28]:
mean_kilos_by_brand = {}

for el in picked_brands:
    mean_kilos = autos[autos['brand'] == el]['odometer_km'].mean()
    mean_kilos_by_brand[el] = mean_kilos
    
print('Average kilometers by each top brand\n', mean_kilos_by_brand)

Average kilometers by each top brand
 {'opel': 128120.90790048015, 'audi': 129283.06551297898, 'mercedes_benz': 131021.18829183049, 'bmw': 132973.79229415217, 'volkswagen': 128463.57409713574, 'ford': 123647.0789203963}


In [29]:
bmp_series = pd.Series(mean_price_by_brand)
bmk_series = pd.Series(mean_kilos_by_brand)

analyze_form = pd.DataFrame(bmp_series, columns=['mean_price'])
analyze_form['mean_kilos'] = bmk_series

analyze_form

mean_price     mean_kilos
audi           9484.499876  129283.065513
bmw            8754.601995  132973.792294
ford           4607.311923  123647.078920
mercedes_benz  8670.394059  131021.188292
opel           3348.936054  128120.907900
volkswagen     5907.612495  128463.574097

The mean kilometers by brand data are pretty similar. The least the Ford has and the most BMW has. Looks like the price does not correlate with the car mileage, neither directly nor inversely.

But three premial brands have a bit more average mileage than other. It may mean that those cars run for a longer time before being sold, than non-premial do. This may indicate a higher quality of the car, or about greater owner's affection for the car.

### Translation of German Words into English

In [30]:
print(autos['vehicle_type'].unique())
translation = {'limousine' : 'limousine',
               'kleinwagen' : 'small_car',
               'kombi' : 'estate_car',
               'bus' : 'bus',
               'cabrio' : 'convertible',
               'coupe' : 'coupe',
               'suv' : 'suv',
               'andere' : 'other',
              }
autos['vehicle_type'] = autos['vehicle_type'].map(translation)
print(autos['vehicle_type'].unique())

['bus' 'limousine' 'kleinwagen' 'kombi' nan 'coupe' 'suv' 'cabrio'
 'andere']
['bus' 'limousine' 'small_car' 'estate_car' nan 'coupe' 'suv'
 'convertible' 'other']


We translated some words from `vehicle_type` column. Let's do the same with `gearbox`, `fuel_type` and `unrepaired damage`.

In [31]:
print(autos['gearbox'].value_counts())
translation = {'manuell' : 'manual',
               'automatik' : 'automatic',
              }
autos['gearbox'] = autos['gearbox'].map(translation)
print(autos['gearbox'].value_counts())

manuell      33367
automatik     9920
Name: gearbox, dtype: int64
manual       33367
automatic     9920
Name: gearbox, dtype: int64


In [32]:
print(autos['fuel_type'].unique())
translation = {'lpg' : 'lpg',
               'benzin' : 'gasoline',
               'diesel' : 'diesel',
               'cng' : 'cng',
               'hybrid' : 'hybrid',
               'elektro' : 'electro',
               'andere' : 'other',
              }
autos['fuel_type'] = autos['fuel_type'].map(translation)
print(autos['fuel_type'].unique())

['lpg' 'benzin' 'diesel' nan 'cng' 'hybrid' 'elektro' 'andere']
['lpg' 'gasoline' 'diesel' nan 'cng' 'hybrid' 'electro' 'other']


In [33]:
print(autos['unrepaired_damage'].unique())
translation = {'ja' : 'yes',
               'nein' : 'no',
              }
autos['unrepaired_damage'] = autos['unrepaired_damage'].map(translation)
print(autos['unrepaired_damage'].unique())

['nein' nan 'ja']
['no' nan 'yes']


### Converting Dates into Pandas Datetime Format

We will convert date from string object to time object. This will be convenient for various trials and further possible research related to time. 

In [34]:
print(autos['date_crawled'].head())
autos['date_crawled'] = pd.to_datetime(autos['date_crawled'])
print(autos['date_crawled'].head())

autos['last_seen'] = pd.to_datetime(autos['last_seen'])
autos['ad_created'] = pd.to_datetime(autos['ad_created'])

0    2016-03-26 17:47:46
1    2016-04-04 13:38:56
2    2016-03-26 18:57:24
3    2016-03-12 16:58:10
4    2016-04-01 14:38:50
Name: date_crawled, dtype: object
0   2016-03-26 17:47:46
1   2016-04-04 13:38:56
2   2016-03-26 18:57:24
3   2016-03-12 16:58:10
4   2016-04-01 14:38:50
Name: date_crawled, dtype: datetime64[ns]


### Determining the Lifetime of Each Ad

Using the `last_seen` and `ad_created` columns we can calculate how long an ad remains online until the car is sold. New column name is `lifetime`.

In [35]:
autos['lifetime'] = autos['last_seen'] - autos['ad_created']

autos['lifetime'].dt.days.value_counts(normalize=True)

0      0.134270
2      0.103363
4      0.074540
1      0.056648
6      0.055695
3      0.050529
8      0.047270
9      0.037470
11     0.034610
5      0.033967
7      0.033834
10     0.026694
13     0.025918
12     0.024854
14     0.023502
15     0.021795
18     0.021218
16     0.019733
17     0.018469
22     0.017294
21     0.015963
19     0.012726
24     0.011928
23     0.011862
29     0.011440
27     0.011396
25     0.010953
26     0.010398
20     0.010199
28     0.008891
         ...   
59     0.000089
34     0.000067
49     0.000067
55     0.000044
43     0.000044
44     0.000044
47     0.000044
42     0.000022
36     0.000022
149    0.000022
53     0.000022
52     0.000022
62     0.000022
51     0.000022
98     0.000022
209    0.000022
48     0.000022
61     0.000022
239    0.000022
68     0.000022
88     0.000022
63     0.000022
109    0.000022
70     0.000022
45     0.000022
39     0.000022
300    0.000022
58     0.000022
41     0.000022
95     0.000022
Name: lifetime, Length: 

As we can see, 13% of ads existed for less than a day. This may be caused by the premature removal of the ad due to errors in the description or another. Let's calculate how many cars were sold in the first two weeks.

In [36]:
(autos[autos['lifetime'].dt.days != 0].loc[:,'lifetime']
.dt.days.value_counts(normalize=True).sort_index() [0:14].sum()
)

0.7264322483161318

So, ~73% of all cars were sold in the first two weeks

### Exploring the Most Common Brand + Model Combination



In [37]:
brand_model = autos['brand'] + ': ' + autos['model']
brand_model.value_counts(normalize=True).head(10)

volkswagen: golf           0.084164
bmw: 3er                   0.060021
volkswagen: polo           0.032903
opel: corsa                0.031184
volkswagen: passat         0.031161
opel: astra                0.029836
audi: a4                   0.028419
mercedes_benz: c_klasse    0.026653
bmw: 5er                   0.026304
mercedes_benz: e_klasse    0.022377
dtype: float64

The most common Brand-Model combination is Volkswagen Golf. It occupies 8% of surveyed market.

### Finding Dependence of the Price on the Mileage

We'll split the `odometer_km` into groups, and use aggregation to see if average prices follows any patterns based on the mileage.

In [76]:
#autos['odometer_km'].describe()
mileage = []
mileage_prices = []
car_number = []

for kms in range(10000, 150001, 20000):
    
    sample = ((autos['odometer_km'] <= kms) 
            & (autos['odometer_km'] > (kms - 20000)))
    sample = autos[sample]
                                              
    mean_price = int(sample['price_$'].mean())
    num = sample.shape[0]
    
    mileage.append(kms)
    mileage_prices.append(mean_price)
    car_number.append(num)

In [78]:
mileage = pd.Series(mileage)
mileage_prices = pd.Series(mileage_prices)
car_number = pd.Series(car_number)

analyze_f = pd.DataFrame(mileage, columns=['mileage'])
analyze_f['mileage_prices'] = mileage_prices
analyze_f['car_number'] = car_number

analyze_f

mileage  mileage_prices  car_number
0    10000           18853         854
1    30000           17855        1492
2    50000           15394        1805
3    70000           11790        2320
4    90000            9176        3061
5   110000            8338        2031
6   130000            6609        4839
7   150000            4166       28701

Logically and mathematically, we see the expected dependence - Price falls with increasing mileage. Dependence is inverse

### Discount for Damage Exploring

How much cheaper are cars with damage than their non-damaged counterparts?

In [90]:
print(round(autos[autos['unrepaired_damage'] == 'yes']
            .loc[:,'price_$'].mean())
                                     , '$ mean price with damage')
print('for', autos[autos['unrepaired_damage'] == 'yes'].shape[0], 'cars')

3031.0 $ mean price with damage
for 3675 cars


In [93]:
print(round(autos[autos['unrepaired_damage'] == 'no']
            .loc[:,'price_$'].mean())
                                     , '$ mean price with no damage')
print('for', autos[autos['unrepaired_damage'] == 'no'].shape[0], 'cars')

7443.0 $ mean price with no damage
for 33909 cars


In [94]:
7443 / 3031

2.4556252062025736

So, cars with damage are commonly 2.5 times cheaper than normal cars without damage.

# Conclusions

* The data has been cleaned and transformed for various further investigations.
* We found top 6 car brands in Germany. Audi, BMW and Mercedes Benz as premial brands. The cheapest is Opel, and the medium are Volkswagen and Ford.
* The mean kilometers by brand are pretty similar. In this price does not correlate with the car mileage. 
* Top premial brands have a bit more average mileage, it may mean those cars run for a longer time before being sold. This may indicate a higher quality of the car, or about greater owner's affection for the car.
* 13% of ads existed for less than a day. Perhaps it caused by the premature removal due to errors in the ad.
* 73% of all cars were sold in the first two weeks.
* The most common Brand-Model is Volkswagen Golf (8% of market).
* Price logically falls with increasing mileage among all cars.
* Cars with damage are commonly 2.5 times cheaper than normal cars.